pkg/didcomm/common/service/service.go

```diff
@@ -30,6 +37,39 @@ type DIDCommMsg struct {
 	Payload  []byte
 	// TODO : might need refactor as per the issue-226
 	OutboundDestination *Destination
+	thID                *string
+}
+
+func (m *DIDCommMsg) parseID() (string, error) {
+	msg := struct {
+		ID     string           `json:"@id"`
+		Thread decorator.Thread `json:"~thread"`
+	}{}
+
+	if err := json.Unmarshal(m.Payload, &msg); err != nil {
+		return "", fmt.Errorf("cannot unmarshal @id and ~thread: error=%v", err)
+	}
+
+	if len(msg.Thread.ID) > 0 {
+		return msg.Thread.ID, nil
+	}
+
+	return msg.ID, nil
+}
+
+// ThreadID returns msg ID
```

Comment:
the function also sets the TreadID in DIDCommMsg .. it doesn't just return msg ID

Reply:
`thID *string` is cache,  I did it to avoid unmarshaling twice, the responsibility of this function is to return an ID and no one should care about what it does under the hood

Reply:
Done!